lib/healthcheck/healthcheck-stats.c

```diff
@@ -0,0 +1,164 @@
+#include "stats/stats-registry.h"
+#include "stats/stats-counter.h"
+
+#include "apphook.h"
+#include "cfg.h"
+#include "mainloop.h"
+#include "timeutils/misc.h"
+
+#include <iv.h>
+
+typedef struct _HealthCheckStats
+{
+  HealthCheckStatsOptions options;
+  struct iv_timer timer;
+
+  StatsCounterItem *io_worker_latency;
+  StatsCounterItem *mainloop_io_worker_roundtrip_latency;
+} HealthCheckStats;
+
+static HealthCheckStats healthcheck_stats;
+
+static void
+healthcheck_stats_update(HealthCheckResult result, gpointer c)
+{
+  HealthCheckStats *self = (HealthCheckStats *) c;
+
+  stats_counter_set(self->io_worker_latency, result.io_worker_latency);
+  stats_counter_set(self->mainloop_io_worker_roundtrip_latency, result.mainloop_io_worker_roundtrip_latency);
+}
+
+static void
+healthcheck_stats_timer_start(HealthCheckStats *self)
+{
+  if (self->options.freq <= 0)
+    return;
+
+  iv_validate_now();
+  self->timer.expires = iv_now;
+  timespec_add_msec(&self->timer.expires, self->options.freq * 1000);
+  iv_timer_register(&self->timer);
+}
+
+static void
+healthcheck_stats_timer_stop(HealthCheckStats *self)
+{
+  if (self->timer.handler && iv_timer_registered(&self->timer))
+    iv_timer_unregister(&self->timer);
+}
+
+static void
+healthcheck_stats_run(gpointer c)
+{
+  HealthCheckStats *self = (HealthCheckStats *) c;
+
+  HealthCheck *hc = healthcheck_new();
+  healthcheck_run(hc, healthcheck_stats_update, self);
+  healthcheck_unref(hc);
+
+  healthcheck_stats_timer_start(self);
+}
+
+static void
+_register_counters(HealthCheckStats *self)
+{
+  StatsClusterKey sc_key_io_worker_latency, sc_key_mainloop_iow_rt_latency;
+  stats_cluster_single_key_set(&sc_key_io_worker_latency, "io_worker_latency_nanoseconds", NULL, 0);
+  stats_cluster_single_key_set(&sc_key_mainloop_iow_rt_latency,
+                               "mainloop_io_worker_roundtrip_latency_nanoseconds", NULL, 0);
+
+  stats_lock();
+  stats_register_counter(1, &sc_key_io_worker_latency, SC_TYPE_SINGLE_VALUE, &self->io_worker_latency);
+  stats_register_counter(1, &sc_key_mainloop_iow_rt_latency, SC_TYPE_SINGLE_VALUE,
+                         &self->mainloop_io_worker_roundtrip_latency);
+  stats_unlock();
+}
+
+static void
+_unregister_counters(HealthCheckStats *self)
+{
+  StatsClusterKey sc_key_io_worker_latency, sc_key_mainloop_iow_rt_latency;
+  stats_cluster_single_key_set(&sc_key_io_worker_latency, "io_worker_latency_nanoseconds", NULL, 0);
+  stats_cluster_single_key_set(&sc_key_mainloop_iow_rt_latency,
+                               "mainloop_io_worker_roundtrip_latency_nanoseconds", NULL, 0);
+
+  stats_lock();
+  stats_unregister_counter(&sc_key_io_worker_latency, SC_TYPE_SINGLE_VALUE, &self->io_worker_latency);
+  stats_unregister_counter(&sc_key_mainloop_iow_rt_latency, SC_TYPE_SINGLE_VALUE,
+                           &self->mainloop_io_worker_roundtrip_latency);
+  stats_unlock();
+}
+
+void
+healthcheck_stats_init(HealthCheckStatsOptions *options)
```

Comment:
I think it would be worth to skip the init and deinit if the config is on an insufficient stats level.

Reply:
Or it could be done in timer_start(), like with freq. :)

Reply:
Or it could be done in timer_start(), like with freq. :)

Reply:
Done. Thanks.